Migrate Nonce test to Playwright

pooja-muchandikar opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
## What?
<!-- In a few words, what is the PR actually doing? -->
Part of https://github.com/WordPress/gutenberg/issues/38851.
Migrate[ nonce.test.js ](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/plugins/nonce.test.js) to its Playwright version.
## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Part of https://github.com/WordPress/gutenberg/issues/38851.
## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
See[ MIGRATION.md](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md) for migration steps.
## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Run` npm run test-e2e:playwright test/e2e/specs/editor/plugins/nonce.spec.js`

pooja-muchandikar committed:
Migrate Nonce test to Playwright

pooja-muchandikar committed:
Update Comments

pooja-muchandikar commented:
Hi @kevin940726, @talldan 👋

I hope you are doing great!

I have raised another migration PR, please could anyone of you help me in reviewing this PR?

Thanks!!

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,31 @@
+		// `saveDraft` waits for saving to be successful, so this test would
+		// timeout if it's not.
+		await page.click( 'role=button[name=/Save draft/i]' );
+		// We expect a 403 status once.
+		page.on( 'console', ( message ) => {
+			expect( message.type() ).toBe( 'error' );
+		} );
```
I wonder if we could use the official [`response`](https://playwright.dev/docs/api/class-request) event for this assertion. Something like:

```js
const saveDraftResponse = await new Promise(resolve => {
  page.on('response', response => {
    if (response.url() === 'the errored url') {
      resolve(response);
    }
  });
});

expect(saveDraftResponse.status()).toBe(403);
```

pooja-muchandikar replied:
Hi @kevin940726,

I tried the above suggestion but somehow the API's response is 200.

![image](https://user-images.githubusercontent.com/41000648/189888432-140196b2-3a94-428c-ae69-47356469606b.png)

Because of which kept the console message as is..

pooja-muchandikar replied:
Hi @kevin940726,
Any further suggestions here?

kevin940726 replied:
The original test here is a bit weird as it doesn't tell which response should error. My suggestion would be to try to figure out which response should error and explicitly test it in this test, then follow the code I wrote above.

kevin940726 replied:
This is tricky. This test depends on the old Jest-Puppeteer behavior to auto re-auth manually, it however is not what the new testing architecture in Playwright does now though. I have a working local refactor that uses `page.route` to mock the API so that we don't need the nonce test plugin and the 5 seconds timeout. I'll share more soon.

pooja-muchandikar replied:
Sure!

kevin940726 replied:
Opened https://github.com/WordPress/gutenberg/pull/44929. It's a bigger change than I thought. I figured opening a new PR might be easier in this case 😅 . Your contribution is still very helpful though! We can co-author it when it's merged ;).

pooja-muchandikar replied:
Sure, I am going through the PR you just created. 

Should I close this PR then?

kevin940726 replied:
Yes, that'd be great :) Thank you!

pooja-muchandikar replied:
Sure.

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,31 @@
+		// eslint-disable-next-line no-restricted-syntax
+		await page.waitForTimeout( 5000 );
+		await page.keyboard.type( 'test' );
+		// `saveDraft` waits for saving to be successful, so this test would
+		// timeout if it's not.
+		await page.click( 'role=button[name=/Save draft/i]' );
```
We're not using the `saveDraft` util so this comment is not relevant anymore.

pooja-muchandikar replied:
Done

pooja-muchandikar committed:
Addressed Feedbacks

pooja-muchandikar committed:
Update comments

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,31 @@
+		// eslint-disable-next-line no-restricted-syntax
+		await page.waitForTimeout( 5000 );
+		await page.keyboard.type( 'test' );
+		// Click on save draft button so that this test would
+		// timeout if it's not.
+		await page.click( 'role=button[name=/Save draft/i]' );
```
The comment still reads a bit weird, and I don't think is relevant anymore.

Let's refactor this into two steps:

```js
// Saving draft should still succeed.
await page.click( 'role=button[name=/Save draft/i]' );
await expect(
  page.locator( 'role=button[name="Dismiss this notice"i]' )
).toContainText( /Draft saved/i );
```

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,31 @@
+		await admin.createNewPost();
+		await page.keyboard.press( 'Enter' );
+		// eslint-disable-next-line no-restricted-syntax
+		await page.waitForTimeout( 5000 );
```
It's a shame that we have to explicitly wait for 5 seconds here and **block** the whole testing process to test this one case. I wonder if we could refactor this with API mocking in the future 🤔.

pooja-muchandikar committed:
Addressed Feedbacks

pooja-muchandikar commented:
Hi @kevin940726,

I have addressed all the feedbacks, please let me know if any further changes are required. 

Thanks!

pooja-muchandikar commented:
Hi @kevin940726,

Any further suggestions on this PR?

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,32 @@
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.describe( 'Nonce', () => {
+	// While using beforeEach/afterEach is suboptimal for multiple tests, they
+	// are used here to ensure that the nonce plugin doesn't interfere with API
+	// calls made in global before/after calls, which may perform API requests.
```
I don't think this comment is relevant anymore, let's remove it?

kevin940726 commented on test/e2e/specs/editor/plugins/nonce.spec.js:
```diff
@@ -0,0 +1,32 @@
+	test( 'should refresh when expired', async ( { page, admin } ) => {
+		await admin.createNewPost();
+		await page.keyboard.press( 'Enter' );
+		// eslint-disable-next-line no-restricted-syntax
```
I think we forgot to delete this comment too 😅 .

```suggestion
```

pooja-muchandikar committed:
Addressed feedbacks

pooja-muchandikar commented:
Hi @kevin940726,

Addressed feedbacks. 

CI is failing but seems like not related to the test case I worked upon. 

![image](https://user-images.githubusercontent.com/41000648/195537512-f1d48bb2-3690-405f-b9ea-c5ed9d92595a.png)

pooja-muchandikar commented:
Closing this PR as new PR is created to handle the tricky condition https://github.com/WordPress/gutenberg/pull/44929 

Thanks @kevin940726 for all the feedbacks and help!...
